Add flag to specify classes are unmanaged by default

mppf opened the pull request:
Adds `--default-unmanaged` as a (currently) internal/developer flag.
It should get more testing somehow if it's to be a user-facing flag.

Resolves #9289.

Passed full local testing.
Reviewed by @vasslitvinov - thanks!

mppf commented:
@vasslitvinov - could you have a look at this one after PR #9612 ? thanks

mppf commented:
@vasslitvinov - this PR is ready for your review, thanks!

vasslitvinov commented on compiler/passes/scopeResolve.cpp:
```diff
@@ -162,6 +165,36 @@ void scopeResolve() {
+          INT_ASSERT(0);
+        }
+      }
+    }
```
Please move the above into a separate function over 'fn', ex. adjustMethodsForDefaultUnmanaged(FnSymbol* fn).

vasslitvinov replied:
While there, how about move 'fDefaultUnmanaged' to be the first expression in the initial if-condition ?
That IMO makes it easier to read and faster to execute in the common case.

vasslitvinov commented on compiler/passes/scopeResolve.cpp:
```diff
@@ -725,43 +758,80 @@ void resolveUnresolvedSymExprs(BaseAST* inAst) {
+          call->isNamed("Shared"));
+}
+
+// Returns the expr resulting, either se or a call containing it
```
What if this function returns the resulting Type* directly?

Either way, I'd change the above comment to clarify the intention. For example:

"Handle the case where se->symbol() is "unstable" i.e. an undecorated class type. With --warn-unstable, issue a warning. With --default-unmanaged, wrap it in a _toUnmanaged call. Return either se or a call _toUnmanaged(se)".

vasslitvinov commented on modules/dists/SparseBlockDist.chpl:
```diff
@@ -125,21 +137,15 @@ class SparseBlockDom: BaseSparseDomImpl {
-    var comp = new TargetLocaleComparator();
+    var comp = new TargetLocaleComparator(rank=rank, idxType=idxType,
+                                          sparseLayoutType=sparseLayoutType,
+                                          dist=dist);
```
If the changes to this file in this PR are intentional, please explain them briefly in the PR message.

mppf commented:
@vasslitvinov - I addressed your feedback and moved the SparseBlock change to #9739. I'm planning to merge once another testing run completes successfully.

vasslitvinov commented:
@mppf - sounds good.
